tests/basic/uncontrolled_randomness.rs

```diff
@@ -0,0 +1,231 @@
+use rand::{rngs::StdRng, SeedableRng};
+use shuttle::rand::Rng;
+use shuttle::scheduler::{DfsScheduler, Scheduler};
+use shuttle::sync::{Arc, Mutex};
+use shuttle::{check_uncontrolled_randomness, thread};
+use test_log::test;
+
+fn check_uncontrolled_randomness_custom_scheduler_and_config<F, S>(f: F, scheduler: S)
+where
+    F: Fn() + Send + Sync + 'static,
+    S: Scheduler + 'static,
+{
+    use shuttle::scheduler::UncontrolledRandomnessCheckScheduler;
+    use shuttle::Config;
+
+    let config = Config::default();
+
+    let scheduler = UncontrolledRandomnessCheckScheduler::new(scheduler);
+
+    let runner = shuttle::Runner::new(scheduler, config);
+    runner.run(f);
+}
+
+fn randomly_acquire_lock<R: rand::RngCore, F: (Fn() -> R) + Send + Sync>(thread_rng: &'static F) {
+    const NUM_THREADS: u32 = 10;
+
+    let lock = Arc::new(Mutex::new(0u64));
+    let threads: Vec<_> = (0..NUM_THREADS)
+        .map(|_| {
+            let my_lock = lock.clone();
+
+            thread::spawn(move || {
+                let x = thread_rng().gen::<u64>();
+
+                if x % 10 == 0 {
+                    let mut num = my_lock.lock().unwrap();
+                    *num += 1;
+                }
+            })
+        })
+        .collect();
+
+    threads.into_iter().for_each(|t| t.join().expect("Failed"));
+}
+
+fn have_n_threads_acquire_mutex(num_threads: u64) {
+    let lock = Arc::new(Mutex::new(0u64));
+    let threads: Vec<_> = (0..num_threads)
+        .map(|_| {
+            let my_lock = lock.clone();
+
+            thread::spawn(move || {
+                let mut num = my_lock.lock().unwrap();
+                *num += 1;
+            })
+        })
+        .collect();
+
+    threads.into_iter().for_each(|t| t.join().expect("Failed"));
+}
+
+fn spawn_random_amount_of_threads<F: (Fn() -> Box<dyn rand::RngCore>) + Send + Sync>(
+    thread_rng: &'static F,
+    max_threads: u64,
+) {
+    let num_threads: u64 = thread_rng().gen::<u64>() % max_threads;
+    have_n_threads_acquire_mutex(num_threads);
+}
+
+fn spawn_random_amount_of_threads_mutex_rng(rng: &Mutex<StdRng>, max_threads: u64) {
+    let num_threads = rng.lock().unwrap().gen::<u64>() % max_threads;
+    have_n_threads_acquire_mutex(num_threads);
+}
+
+fn have_n_threads_yield(num_threads: u64) {
```

Comment:
Could you move these helper functions down to just before the tests that use them?  It's much easier to read code when definitions are only introduced before their first use.